### ISSUE TITLE
[BUG] no local data file found after 5 tries - is this a first time run?

### ISSUE THREAD

**Issue by jamespreedy** [wills106/homeassistant-solax-modbus#445]
**Describe the bug**
On Home Assistant boot I get a warning in the logs saying ...
"no local data file found after 5 tries - is this a first time run?"

**Mandatory details**
1. Integration Version: 2023.06.8
2. HA Version: 2023.6.2
3. Inverter Brand: SolaX
4. Plugin Used: PocketLAN
5. First 6 digits of Serial: H34A10
6. Connection Method: LAN / MODBUS

**Detailed Error Log**
2023-06-15 09:24:08.139 WARNING (MainThread) [custom_components.solax_modbus] no local data file found after 5 tries - is this a first time run ?

**Additional context**
Doesn't seem to be impacting anything, but then I'm likely missing something in the background.
Just thought I'd let you know.

If you need any more information, please let me know.
Thanks in advance for your help. 😀

**Comment by infradom** [comment]
This warning is normal during the first run, as you probably have other errors (inverter not recognized?) the system will not be able to write the local data, so I suggest to look into the other error messages first. These warnings will disappear when communication is ok

**Comment by jamespreedy** [comment]
Thanks. Interestingly, no other SolaX errors, just this one. Are there any other things I could check with this one?

**Comment by infradom** [comment]
> Thanks. Interestingly, no other SolaX errors, just this one. Are there any other things I could check with this one?

Is your integration showing all the entities (maybe over 150 for a Gen4) ? Or do you only see a couple of entities ?
If you only see a couple of entities, it is probably caused by the pocketlan interface that does not support Modbus over TCP. Most people use an external Ethernet to Modbus adapter (see Wiki)
Are you sure this warning is still coming after a restart ?
In case you see all the >150  entities, we are writing some variables to a file  config/SolaX_data.json after first startup.

**Comment by jamespreedy** [comment]
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/559531/52386754-2c65-47c0-a944-7c254702759a)


Seem to have plenty there. As I say, everything else seems ok other than this odd error. Is there a way I could delete that file and give it a chance to start up fresh?

Thanks again!

**Comment by infradom** [comment]
I guess that the system was not able to create that file for some unclear
reason. What type of HA installation do you run?

Op do 15 jun. 2023 19:05 schreef James Preedy ***@***.***>:

**Comment by jamespreedy** [comment]
Hey, thanks for your help. I'm running Home Assistant OS on an Intel Nuc.
Everything's up to date.

**Comment by jamespreedy** [comment]
OK, so I logged in over ssh and created the /root/config/SolaX_data.json file just in case.
I rebooted the host rather than just Home Assistant, and I now get the below error.

I've not had any issues in the past, or changed anything recently other than kept my system up to date.
The 254x entities seem fine, and are updating regularly.


```
This error originated from a custom integration.

Logger: custom_components.solax_modbus
Source: custom_components/solax_modbus/__init__.py:249 
Integration: SolaX Inverter Modbus (documentation, issues) 
First occurred: 9:46:37 pm (1 occurrences) 
Last logged: 9:46:37 pm

Something went wrong reading from modbus
Traceback (most recent call last):
  File "/config/custom_components/solax_modbus/__init__.py", line 428, in read_modbus_data
    res = self.read_modbus_registers_all()
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/solax_modbus/__init__.py", line 531, in read_modbus_registers_all
    if not self.localsLoaded: self.loadLocalData()
                              ^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/solax_modbus/__init__.py", line 249, in loadLocalData
    loaded = json.load(fp)
             ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/__init__.py", line 293, in load
    return loads(fp.read(),
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

```

**Comment by jamespreedy** [comment]
I've deleted the /root/config/SolaX_data.json file again (it was still empty) and rebooted.
I'm back to the original error.

`2023-06-15 09:24:08.139 WARNING (MainThread) [custom_components.solax_modbus] no local data file found after 5 tries - is this a first time run ?`

**Comment by infradom** [comment]
Thanks for these tests.

I reaxamined the code and maybe the file will only be written if one of the LOCAL_DATA  parameters is modified.
Can you try to modify remotecontrol_duration to a different value (eg 15 instead of 20 ?) (or modify any other remotecontrol_xxx entitiy)
This will probably create the file and the message will disappear on next reboot.

We will try to correct this error, but if the problem disappears after this small test, the warning message is innocent ...

**Comment by jamespreedy** [comment]
This fixed it immediately - thanks. As soon as I changed remotecontrol_duration, I saw that the file was created.
On rebooting, no more errors. Yay!

Thank you so much!! 😀

**Comment by wills106** [comment]
Closing down as this was addressed in [2023.07.1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.07.1)
